packages/core/src/sync-realtime/service.ts

```diff
@@ -8,878 +8,633 @@ import {
+ * compatible with out local chain.
+ *
+ * @param pendingLatestBlock Block that caused reorg to be detected.
+ * Must be at most 1 block ahead of the local chain.
+ */
+export const handleReorg = async (
+  realtimeSyncService: Service,
+  pendingLatestBlock: SyncBlock,
+) => {
+  // Prune the local chain of blocks that have been reorged out
+  realtimeSyncService.localChain = realtimeSyncService.localChain.filter(
+    ({ block }) => block.number < hexToNumber(pendingLatestBlock.number),
+  );
+
+  // Block we are attempting to fit into the local chain.
+  let reorgedBlock = pendingLatestBlock;
+
+  while (true) {
+    const parentBlock = getLatestLocalBlock(realtimeSyncService);
+
+    if (parentBlock.hash === reorgedBlock.parentHash) {
+      realtimeSyncService.common.logger.trace({
+        service: "realtime",
+        msg: `Found common ancestor block ${parentBlock.number} (network=${realtimeSyncService.network.name})`,
+      });
 
-      await this.syncStore.deleteRealtimeData({
-        chainId: this.network.chainId,
-        fromBlock: BigInt(commonAncestor.number),
+      await realtimeSyncService.syncStore.deleteRealtimeData({
+        chainId: realtimeSyncService.network.chainId,
+        fromBlock: BigInt(parentBlock.number),
       });
 
-      this.emit("reorg", {
-        ...maxCheckpoint,
-        blockTimestamp: commonAncestor.timestamp,
-        chainId: this.network.chainId,
-        blockNumber: commonAncestor.number,
+      realtimeSyncService.onEvent({
+        type: "reorg",
+        chainId: realtimeSyncService.network.chainId,
+        safeCheckpoint: {
+          ...maxCheckpoint,
+          blockTimestamp: parentBlock.timestamp,
+          chainId: realtimeSyncService.network.chainId,
+          blockNumber: parentBlock.number,
+        },
       });
 
-      this.common.logger.warn({
+      realtimeSyncService.common.logger.warn({
         service: "realtime",
-        msg: `Detected invalid logs returned by the RPC starting at block ${commonAncestor.number} (network=${this.network.name})`,
+        msg: `Detected reorg with common ancestor ${parentBlock.number} (network=${realtimeSyncService.network.name})`,
       });
-    };
 
-    for (let i = 0; i < localLogs.length && i < matchedLogs.length; i++) {
-      const lightMatchedLog = realtimeLogToLightLog(matchedLogs[i]);
-      if (
-        lightMatchedLog.blockHash !== localLogs[i].blockHash ||
-        lightMatchedLog.logIndex !== localLogs[i].logIndex
-      ) {
-        await handleLogsInconsistency(localLogs[i].blockNumber - 1);
-        return true;
-      }
+      return;
     }
 
-    if (localLogs.length !== matchedLogs.length) {
-      await handleLogsInconsistency(
-        localLogs.length === 0
-          ? undefined
-          : localLogs[localLogs.length - 1].blockNumber - 1,
-      );
-      return true;
+    if (realtimeSyncService.localChain.length === 0) break;
+    else {
+      reorgedBlock = await _eth_getBlockByHash(realtimeSyncService, {
```

Comment:
Previously, we had a little optimization where we'd hold onto these remote blocks as we fetch them, and then add them to the queue once we find a common ancestor. I'm fine to leave that out if you think it might be bug-prone, just thought I'd mention it.

Reply:
I don't think the savings are worth the complexity addition